### PR TITLE
Storing signed vote requests in a new table (votinghistory)

### PR DESCRIPTION
**Pull request by rschust**

### CODE REVIEW

File: src/node/entities.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ namespace ccf
     static constexpr auto APP_SCRIPTS = "appscripts";
     static constexpr auto SECRETS = "secrets";
     static constexpr auto CODEID = "codeid";
+    static constexpr auto VOTING_HISTORY = "voting_history";
```
Comment (jumaffre) [inline src/node/entities.h]: For consistency, we should probably remove the underscore for now.

**Comment by achamayou** [comment]
@rschust as discussed, let's add an end to end test consisting of a round of voting, and an offline check of the member votes.

File: tests/ledger.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
+import io
```
Comment (jumaffre) [inline tests/ledger.py]: Copyright missing

File: tests/ledger.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        self._file.seek(0, 2)
+        self._file_size = self._file.tell()
+        self._file.seek(0, 0)
+
+    def __del__(self):
+        self._file.close()
+
+    def read_header(self):
+        # read the size of the transaction
+        buffer = _byte_read_safe(self._file, 4)
```
Comment (jumaffre) [inline tests/ledger.py]: Could we define `4` (and `8` on line `116`) (like `GCM_SIZE_TAG`)?

**Comment by jumaffre** [comment]
We should also update https://microsoft.github.io/CCF/kv/kv_serialisation.html and explain how the new python utility can be used to read/verify the ledger.

File: tests/ledger.py (hunk 2 above)
Comment (achamayou) [reply tests/ledger.py]: The script should catch that, if not, it needs the re-configured to pick up this file too.

File: src/node/rpc/memberfrontend.h [hunk 4]
```diff
@@ -301,6 +301,14 @@ namespace ccf
         // record vote
         proposal->votes[args.caller_id] = vote.ballot;
         proposals->put(vote.id, *proposal);
+
+        // record signed vote request
+        if (!args.signed_request.sig.empty())
```
Comment (jumaffre) [inline src/node/rpc/memberfrontend.h]: Should the command fail if there is not signature for this request?

File: tests/votinghistory.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+    ledger_filename = None
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        # create a lua query file to change a member state to accepted
+        with open("query.lua", "w") as qfile:
```
Comment (jumaffre) [inline tests/votinghistory.py]: Could we re-use the script in `memberclient.py` instead?

File: tests/votinghistory.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+        # create json file to be passed as the argument to the query.lua file
+        # it is passing a member id
+        with open("param.json", "w") as pfile:
+            pfile.write("""{"p": 0}""")
+
+        # propose to add a new member
+        # proposal number 0
+        infra.proc.ccall("./genesisgenerator", "cert", "--name=member4")
+        result = infra.proc.ccall(
```
Comment (jumaffre) [inline tests/votinghistory.py]: I wonder if we should extend the `ccf` module to include native python client code for the member RPCs as they are rather verbose?

File: tests/votinghistory.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    l = ledger.Ledger(ledger_filename)
+
+    # this maps a member_id to a cert object, and is updated when we iterate the transactions,
+    # so that we always have the correct cert for a member on a given transaction
+    members = {}
+    verified_votes = 0
+    for tr in l:
+        tables = tr.get_public_domain().get_tables()
+        members_table = tables["membercerts"]
+        for cert, member_id in members_table.items():
```
Comment (jumaffre) [inline tests/votinghistory.py]: Should we only do this after the condition on line `139`?

File: tests/votinghistory.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        j_result = json.loads(result.stdout)
+        assert j_result["result"]
+
+        ledger_filename = network.find_leader()[0].remote.get_ledger_full_path()
+
+    l = ledger.Ledger(ledger_filename)
+
+    # this maps a member_id to a cert object, and is updated when we iterate the transactions,
+    # so that we always have the correct cert for a member on a given transaction
+    members = {}
```
Comment (jumaffre) [inline tests/votinghistory.py]: I wonder if we should create a new python module that knows about CCF's tables, is initialised with a `Ledger` and can "smart" operations on the ledger, e.g. counting verified votes.

File: tests/votinghistory.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            for member_id, signed_request in votinghistory_table.items():
+                # if the signed vote is stored - there has to be a member at this point
+                assert member_id in members
+                cert = members[member_id]
+                sig = signed_request[0][0]
+                req = signed_request[0][1]
+                verify_sig(cert, sig, req)
+                ++verified_votes
+
+    assert verified_votes == 0
```
Comment (jumaffre) [inline tests/votinghistory.py]: Should this be `2`?

File: tests/votinghistory.py (hunk 7 above)
Comment (rschust) [reply tests/votinghistory.py]: Member certs may be added on earlier transactions than the ones that contain the voting, so we need to store them in order to be able to query at a later stage.

File: tests/votinghistory.py (hunk 9 above)
Comment (rschust) [reply tests/votinghistory.py]: Thanks for noticing, it should be 3

File: tests/votinghistory.py (hunk 8 above)
Comment (rschust) [reply tests/votinghistory.py]: This is a good idea, but it means that we would need to maintain the table structures on python as well (and keep them updated whenever we update the c++ code). @achamayou did say that we should strongly consider using a DDL to represent the tables, so I believe there will be a discussion around that, but not as a part of this PR.

File: src/node/rpc/memberfrontend.h (hunk 4 above)
Comment (rschust) [reply src/node/rpc/memberfrontend.h]: We do not enforce this for any of the other RPC requests, maybe we should discuss if there should be a list of request types that are required to be signed.

File: tests/votinghistory.py (hunk 5 above)
Comment (rschust) [reply tests/votinghistory.py]: Sorry, this shouldn't be here at all.

File: tests/votinghistory.py (hunk 6 above)
Comment (rschust) [reply tests/votinghistory.py]: Sounds reasonable, but it would not be in the scope of this PR.

**Pull request merged**